### ISSUE TITLE
How to read FloatArrayPacket from output stream

### ISSUE THREAD

**Issue by patrick508** [homuler/MediaPipeUnityPlugin#750]
### Plugin Version or Commit ID

v0.10.1

### Unity Version

2021.3.8f1

### Your Host OS

Windows 11

### Target Platform

iOS

### Description

For our Unity project we use this plugin to find the pose on an IOS device. For this we used your sample of the GraphRunner.cs. to add a new stream for our own Tflite Machine Learning model. We know this model gives correct output, but we have trouble getting this output in the format we want float array.

When starting the mediapipe streams, we also create our own stream and in the same way as your samples, we request the output of this stream. This outputs a ImageFrame which in the editor(using the cpu graph) gives us an ImageFrame with format: Vec32F1. If we then call: "streamoutput.TryReadChannel()" we can pass along a float array, which get's filled and this is exactly the data we need for later post processing.

On IOS however, using the gpu graph, this gives us an ImageFrame with format Srgba. The TryReadChannel function that returns a float array only works if the image type is of type float(Vec32F1) so for IOS we are looking at the best way to get the output of the stream, as a float array.

We tried a couple of things:

1. First off we call a different TryReadChannel function in your ImageFrameExtension.cs that is suitable for Srgba type. This returns a byte array, which we then try to convert to a float array, but that messes up the data.
2. Instead of TryReadChannel, I tried the TryReadPixelData() call. this returns a Color32[] which I then convert to a Texture2D, from which I then extract the byte array and finally I convert that to a float array I send to Post Processing. This gives us output, but the output we get with that on IOS, is widely different from what I get inside Unity with the TryReadChannel float call. And above all, this seems like a highly inefficient way to get the float array data we want.
3. Lastly, we've been trying to fix this in the gpu graph itself. Instead of making a output stream that outputs a ImageFrame, we are trying to make a output stream that outputs an FloatArrayPacket, which then should already be a float array we can directly use for post processing. Trying this, we walk into the following issue:

**The relevant part of the graph looks as follows:**
`node {
  calculator: "TfLiteCustomOpResolverCalculator"
  output_side_packet: "OP_RESOLVER:op_resolver"
}
node {
  calculator: "InferenceCalculator"
  input_stream: "TENSORS:input_tensors"
  output_stream: "TENSORS:output_tensors"
  input_side_packet: "MODEL:model"
  input_side_packet: "OP_RESOLVER:op_resolver"
  options: {
    [mediapipe.InferenceCalculatorOptions.ext] {
      delegate {
        xnnpack {}
      }  
    }
  }
}
node: {
    calculator: "TensorsToFloatsCalculator"
    input_stream: "TENSORS:output_tensors"
    output_stream: "FLOATS:segmentation_mask"
}`

**Creating the output stream, and getting the output looks as follows:**

`CalculatorGraph graph = new CalculatorGraph(graphText);
OutputStream<FloatArrayPacket, float[]> segmentationMaskStream = new OutputStream<FloatArrayPacket, float[]>(graph, SEGMENTATION_MASK_OUTPUT_STREAM);
segmentationMaskStream.StartPolling().AssertOk();
stopwatch = new Stopwatch();
stopwatch.Start();
graph.StartRun(sidePacket).AssertOk();
ImageFrame imageFrame = new ImageFrame(ImageFormat.Types.Format.Srgba, inputTexture.width, inputTexture.height,
    inputTexture.width * 4, inputTexture.GetRawTextureData<byte>());
long currentTimestamp = stopwatch.ElapsedTicks / (TimeSpan.TicksPerMillisecond / 1000);
Timestamp timestamp = new Timestamp(currentTimestamp);
graph.AddPacketToInputStream("input_video", new ImageFramePacket(imageFrame, timestamp))
    .AssertOk();
bool segmentationMaskOutput = segmentationMaskStream.TryGetNext(out float[] segmentationMask);`

**Running the above for both cpu & gpu graphs and testing inside the editor gives us the following error:**
`InvalidOperationException: The array's length is unknown, set Length first
Mediapipe.FloatArrayPacket.Get () (at Packages/com.github.homuler.mediapipe/Runtime/Scripts/Framework/Packet/FloatArrayPacket.cs:63)
Mediapipe.Unity.OutputStream2[TPacket,TValue].TryGetNext (TValue& value, System.Int64 timestampThreshold, System.Boolean allowBlock) (at Packages/com.github.homuler.mediapipe/Runtime/Scripts/Unity/OutputStream.cs:229)
Mediapipe.Unity.OutputStream2[TPacket,TValue].TryGetNext (TValue& value, System.Boolean allowBlock) (at Packages/com.github.homuler.mediapipe/Runtime/Scripts/Unity/OutputStream.cs:235)`

To sum up, the question is if our approach of the FloatArrayPacket is the correct way of getting the float[] data we want. If yes, what are we doing wrong. If not, do you have any suggestion on how to get the float[] data we need, while keeping performance in mind(Like not converting to texture2D's).

Thanks for any help in advance!

Patrick

**Comment by homuler** [comment]
> To sum up, the question is if our approach of the FloatArrayPacket is the correct way of getting the float[] data we want.

First, let me answer this question.
`FloatArrayPacket` stores `float[]` internally (in C++), but `TensorsToFloatsCalculator` outputs `std::vector<float>`, so you cannot use `FloatArrayPacket` to receive the output (the error is caused by another issue here, though).
See https://github.com/homuler/MediaPipeUnityPlugin/issues/632#issuecomment-1181729986 and #656 for more details.

> On IOS however, using the gpu graph, this gives us an ImageFrame with format Srgba

> First off we call a different TryReadChannel function in your ImageFrameExtension.cs that is suitable for Srgba type. This returns a byte array, which we then try to convert to a float array, but that messes up the data.

I don't see details in the description, but am I correct in assuming that you are trying to replace the model for the Pose/Selfie Segmentaion?

Depending on the output format, if it is the same as the official Selfie Segmentation, the output value would be between 0 and 255 and be set to the alpha channel.
If so, you can use `TryReadChannelNormalized`.
https://github.com/homuler/MediaPipeUnityPlugin/blob/f196f4631f81afb2f897116c5adbfead3768aab9/Packages/com.github.homuler.mediapipe/Runtime/Scripts/Unity/Extension/ImageFrameExtension.cs#L142-L161
See also the `MaskAnnotationController`.
https://github.com/homuler/MediaPipeUnityPlugin/blob/f196f4631f81afb2f897116c5adbfead3768aab9/Packages/com.github.homuler.mediapipe/Runtime/Scripts/Unity/Annotation/MaskAnnotationController.cs#L43

**Comment by patrick508** [comment]
Thanks or the reply!

To answer your question, we are not trying to replace the model for the Pose/Selfie segmentation. We run an implementation based on your samples, and use the PoseLandmark and PoseWorldLandmarks streams, but on top of that we created our own Tflite model that outputs our own segmentation mask. So we run 3 streams.

![image](https://user-images.githubusercontent.com/9352192/191530448-1851714d-bc52-4ae6-95a5-bd1e3692c007.png)

![image](https://user-images.githubusercontent.com/9352192/191532551-05e03c64-4a08-4c35-ac41-131be19c24b3.png)

To give you a bit more background info, this model we added ourself, outputs a binary mask, for which we then loop through the data and find coordinates that are above a certain threshold (lets say 0.5f), they are a possibility of being our searched for landmark. Each coordinate above this threshold we give a a value of 255, and below get's a value of 0. We then put this inside an OpenCV Mat, and finally do a FindContour to basically get a rough contour of where the landmark is. We then pick the middle coordinate and point that out as our landmark.


Just minutes before you posted a reply we came across the same example in the `MaskAnnotationController`. We just tried this and this gives better output for sure! It still works better in the editor then on IOS, but this might just be our model. So for now I would say this is fixed. Stupid how small of a change this is. We tried literally everything and every combination except TryReadChannelNormalized. Thanks for the reply, much appreciated!
